### ISSUE TITLE
Driver hangs on COPY of wide rows

### ISSUE THREAD

**Issue by levonkorganyan** [impossibl/pgjdbc-ng#488]
Hello,
I've been benchmarking the performance of `COPY TO` between various drivers and psql when I came across what I believe to be a bug with the ng driver. Rows >100MB seem to result in the driver hanging indefinitely.

I used a Postgres test database in the AWS Oregon region with instance sizes t2.medium and m4-xlarge (which has higher bandwidth), and the results were the same.

Here are some results from the benchmark:
|row count x column size|speed (MBps)|
|-|-|
|10240x10KB|27.825012|
|100x1MB|23.458912|
|10x10MB|22.588661|
|1x100MB|N/A|

Script is provided below with some boilerplate for prepping the table. It's the 1 row, 100MB value column case that's hanging.
```java
package com.fivetran.scripts;

import com.impossibl.postgres.api.jdbc.PGConnection;
import com.impossibl.postgres.jdbc.PGDataSource;
import java.io.IOException;
import java.io.OutputStream;
import java.sql.Connection;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

public class NgBenchmark {
    public static void main(String[] args) throws SQLException, IOException {
        new NgBenchmark().run();
    }

    private static final long ROW_BATCH_SIZE = 1000;
    private final PGDataSource source = source();

    private static PGDataSource source() {
        PGDataSource source = new PGDataSource();
        source.setServerName("HOST");
        source.setUser("USER");
        source.setPassword("PASSWORD");
        return source;
    }

    void run() throws SQLException, IOException {
        long rowCount = 1;
        Map<String, Long> columns = new HashMap<>();
        columns.put("a_column", 100 * 1024L * 1024L);

        String tableName =
                "rows_"
                        + rowCount
                        + "_columns_"
                        + columns.values().stream().map(Object::toString).collect(Collectors.joining("_"));

        // prepare
        if (!tableExists(tableName)) {
            System.out.println("Creating and filling table " + tableName);
            createTable(tableName, columns);
            fillTable(tableName, rowCount, columns);
        }

        // copy
        String query = "COPY (SELECT * FROM " + tableName + ") TO STDOUT WITH CSV NULL 'null-lLiTPx0j' ENCODING 'utf8'";
        long copyDuration;
        long bytesCopied;
        try (Connection connection = source.getConnection();
                CountingOutputStream os = new CountingOutputStream()) {
            PGConnection pgConnection = connection.unwrap(PGConnection.class);
            long before = System.currentTimeMillis();
            System.out.println("Starting copy");
            pgConnection.copyTo(query, os);
            copyDuration = System.currentTimeMillis() - before;
            bytesCopied = os.count();
        }

        float durationInSeconds = copyDuration / 1000f;
        float mbCopied = bytesCopied / 1024f / 1024f;
        System.out.println((mbCopied / durationInSeconds) + " MBps");
    }

    private static class CountingOutputStream extends OutputStream {
        private long count;

        @Override
        public void write(int b) throws IOException {
            count += 1;
        }

        public long count() {
            return count;
        }
    }

    void createTable(String tableName, Map<String, Long> columns) throws SQLException {
        String columnClause =
                "("
                        + columns.keySet().stream().map(aLong -> aLong + " VARCHAR").collect(Collectors.joining(", "))
                        + ")";
        execute("CREATE TABLE IF NOT EXISTS " + tableName + " " + columnClause);
    }

    boolean tableExists(String tableName) throws SQLException {
        try (Connection connection = source.getConnection();
                Statement statement = connection.createStatement()) {
            ResultSet resultSet =
                    statement.executeQuery(
                            "SELECT EXISTS ("
                                    + "   SELECT FROM information_schema.tables "
                                    + "   WHERE  table_schema = 'public'"
                                    + "   AND    table_name   = '"
                                    + tableName
                                    + "'"
                                    + "   )");
            resultSet.next();
            return resultSet.getBoolean(1);
        }
    }

    void execute(String sql) throws SQLException {
        try (Connection connection = source.getConnection();
                Statement statement = connection.createStatement()) {
            statement.execute(sql);
        }
    }

    private String junkString(long bytes) {
        long c = 0;
        StringBuilder result = new StringBuilder();
        while (c < bytes) {
            String item = UUID.randomUUID().toString();
            long potentialSize = c + item.length();
            if (potentialSize > bytes) {
                item = item.substring(0, item.length() - ((int) (potentialSize - bytes)));
            }
            c += item.length();
            result.append(item);
        }
        return result.toString();
    }

    String batchClause(long rows) {
        return "generate_series(1, " + rows + ")";
    }

    void fillTable(String tableName, long rowCount, Map<String, Long> columns) throws SQLException {
        Map<String, String> junkStrings = new HashMap<>();
        columns.forEach((key, value) -> junkStrings.put(key, junkString(value)));

        String baseQuery =
                "INSERT INTO "
                        + tableName
                        + " ("
                        + String.join(", ", columns.keySet())
                        + ") SELECT "
                        + columns.keySet()
                                .stream()
                                .map(k -> "'" + junkStrings.get(k) + "'")
                                .collect(Collectors.joining(", "))
                        + " FROM ";

        int i = 0;
        while (i < rowCount) {
            long step = Math.min(i + ROW_BATCH_SIZE, rowCount);
            String fullQuery = baseQuery + batchClause(step);
            execute(fullQuery);
            i += step;
        }
    }
}
```

Please let me know if I can provide any more information that might help. Thank you.

**Comment by kdubb** [comment]
@levonkorganyan This may be the same issue as #489. The `protocol.message.size.max` is set to 20mb as the default.  Can you verify?

**Comment by levonkorganyan** [comment]
@kdubb Just verified that the query is still hanging even after adding the following, unfortunately.
```java
source.setProtocolMessageSizeMax(200 * 1_024 * 1_024);
```

**Comment by kdubb** [comment]
@levonkorganyan Can you validate this with the latest driver?  Changing the `source()` method to the version below allows it to succeed, and you seem to be getting a different result.

```

  private static PGDataSource source() {
    PGDataSource source = new PGDataSource();
    source.setDatabaseName("test");
    source.setServerName("localhost");
    source.setUser("test");
    source.setPassword("test");
    source.setProtocolMessageSizeMax(200 * 1024 * 1024);
    return source;
  }
```

**Note** The _hang_, caused by an error during a copy command, has been fixed in #504

**Comment by kdubb** [comment]
@levonkorganyan I am closing as fixed... feel free to report your finding and re-open if necessary.
